### ISSUE TITLE
Provide a jinja2 template for the kubevirt manifest

### ISSUE THREAD

**Issue by dcritch** [kubevirt/kubevirt#966]
A couple of related projects (kubevirt-ansible, kubevirt apb, kubvirt helm chart) are looking to get a reliable template of the kubevirt manifest, with variables for docker tag, namespace, etc. 

Any chance kubevirt could maintain an official jinja2 version of the manifest that we could all reference? 

cc @fabiand @cynepco3hahue @rthallisey

**Comment by rthallisey** [comment]
@davidvossel and @rmohr do folks have any thoughts on this?

**Comment by davidvossel** [comment]
My concern with this is if we created a jinja2 template for kubevirt.yaml, we can't guarantee it will work with all present/past/future kubevirt releases. The template itself would be specific to a kubevirt release, which then negates the benefit of having a template at all.

Here's an example of what I mean.

I recently added something called a CRD validation webhook into the virt-api component. That feature requires new RBAC permissions for the virt-api service account. These RBAC roles are in the kubevirt.yaml manifest as well.  This means that if we attempted to use a template to run the new version of virt-api, and that template didn't also have the RBAC permission update, then virt-api would fail to start. 

That kind of coupling happens all the time in lots of different ways. Another example is some kubevirt versions expect cli arguments for components that other kubevirt versions wouldn't have.

**Comment by rmohr** [comment]
So should this image simply be there to have a unified way to get a complete yaml of the `current` kubevirt version, where it is easy to exchange `namespace`, `tag` and `registry`? For that, I guess we could provide a template.

It wuld not work for more because of the reasons which @davidvossel mentioned.

**Comment by dcritch** [comment]
That's is about all we are looking for (ns, tag, registry). Some folks might host images on their own internal/disconnected registry. And being able to test different tags benefits CI/CD and upgrades.

Would there be a 'range' of tags that would work for a given release? As per @davidvossel's comment, a particular released manifest may introduce a new CRD, but that should be valid/static over a range of image versions.

@rmohr, so how much of a hassle would it be for kubevirt to maintain that version of the template? I think that would suit our needs.

**Comment by fabiand** [comment]
Couldn't we generate a template (with the parameters mentioned oin the prev comment) which we use to generate our final yaml, and then - in addition to our final yaml - also provide the template (plus, as comments the values for the parameters we used).

T me this has the benefit, that somebody can take the template, with the default values, but i.e. if a custom registry, or a derived build of a container is used, they are free to override that parameter.

In general I agree that manifests are part of a release - aka versioned - as well.

**Comment by rmohr** [comment]
> Would there be a 'range' of tags that would work for a given release? As per @davidvossel's comment, a particular released manifest may introduce a new CRD, but that should be valid/static over a range of image versions.

That is not something which we can guarantee. For updating and deprovisioning KubeVirt, it will never be sufficient to rely on the manifest template. Also not all of our k8s resources which we add to the cluster are part of the manifest.

What we can guarantee:
 * Using such a templatized manifest with the matching versions will successfully deploy a fresh KubeVirt installation
 * All resources which we create will have a `kubevirt.io` label.

What we can **not** guarantee:
 * That all resources are in the manifest
 * That you simply apply a new manifest for updating (if we (re)move components, they may not be deleted)
 * That you simply delete all resources annotated with the `kubevirt.io` labels.

Here the script which we use to remove KubeVirt when switching between branches and syncing the cluster: https://github.com/kubevirt/kubevirt/blob/master/cluster/clean.sh.

What we might be able to do: Introducing lables to resources which should be deleted on an update. But there would still be a gap left.

**Comment by dcritch** [comment]
Ok @rhohr, those caveats are fine. 

We also pull in another manifest, [from containerized-data-importer]](https://github.com/kubevirt/containerized-data-importer/blob/master/manifests/controller/cdi-controller-deployment.yaml). It would be awesome if we had a jinja2'ed version of that as well.

Similar to internal/different registries, some folks might also want to control the registry namespace, e.g. kubevirt vs. openshift3. 

In summary, could we get jinja2 template of the manifest, exposing:
1/ OpenShift/Kubernetes namespace
2/ Registry URL
3/ Registry namespace
4/ docker tag

If everyone is good with this approach, I can open another issue under [containerized-data-importer](https://github.com/kubevirt/containerized-data-importer) if necessary.

Thanks!

**Comment by rmohr** [comment]
That would work for me. @davidvossel?

**Comment by davidvossel** [comment]
> That would work for me. @davidvossel?

I'm fine with this as long as the template is versioned and our automation tools are pulling in the jinja2 template for the specific kubevirt version being installed.

**Comment by lukas-bednar** [comment]
Could we also have templates for `manifests/testing/*.yaml.in`? It would simplify our workflow to run tests.

In order to get these manifests applied on our testing cluster, every time we have to clone kubevirt repository and run `make manifests` to generate these testing manifests which are needed as a prerequisite to run tests.
And this operation can take time on clean system, because it has to build container for building kubevirt.

**Comment by fabiand** [comment]
@lukas-bednar do you need the testing manifests for the master branch or just for releases.

For master would mean that we need to keep the compiled manifests in the repo.
For releases it would mean that we provide the ocmpiled manifests in the releases page.

**Comment by lukas-bednar** [comment]
@fabiand I believe that per release is enough for us, since that is what we consume.

**Comment by rmohr** [comment]
Since the testing manifests need to align with our testing binary, I would prefer to add them as go-structs to our code base (like we were doing with e.g. the examples). 
 * For our normal development flow, we would still just generate yamls and always deploy them
 * For running the tests on external cluster just with our binary, I would prefer a switch in the test binary which can deploy these structs for you.

@lukas-bednar btw. how does getting the testing manifests without running `make generate` help you? Wouldn't you still need to build the testing binary anway? If you just use them for your own testing, then I would suggest to copy the manifests and adjust to your needs and not depend on the files in this repo.

**Comment by lukas-bednar** [comment]
@rmohr hm, that is good point. At the end I will have to build test binary anyway ...

My use-case for these tests jinja templates, is that I have jenkins job which builds cluster for user.
I have two types of users now: 1) ci-trigger, 2) member of our team.
In case of 1) I build test binary and run tests there - no problem with current flow.
In case of 2) ppl are using these clusters for manual testing or as a playground when writing new tests - and in this case I would like to populate these testing manifests for them.

Maybe it is time to revive my #931 . If we have all tests-manifests as structs, then I don't have to unmarshal that yaml files, and I don't need to guess what api to use for creating them - which should simplify that PR.

**Comment by dcritch** [comment]
I'd also vote for having it on the release page.

I dug around to see how the go templates are generated since there is no direct kubevirt.yaml in the repo. I want to follow the same process for the jinja2 template.

I ran that `make generate` but that only seems to make the vm*/ovm manifests. Could someone help me out with how we go from code in the repo to a generated kubevirt.yaml on the releases page?

**Comment by rmohr** [comment]
> I ran that make generate but that only seems to make the vm*/ovm manifests. Could someone help me out with how we go from code in the repo to a generated kubevirt.yaml on the releases page?

#1116 should do that.

**Comment by dcritch** [comment]
So I'm off the hook? :)

That PR looks awesome. The only thing is we also want to be able to specify the registry URL for private/local registries. I suppose we can add that to the `{{ docker_prefix }}`. Or do you have plans to handle that to the Go template? Is it 'legal' or bad form for us to overload `docker_prefix` like that?

**Comment by rmohr** [comment]
> Is it 'legal' or bad form for us to overload docker_prefix like that?

Using it for pointing to different `orgs` or `registries` is exactly what it is for! :+1:

**Comment by fabiand** [comment]
Just to be clear and level expectations: (jinja) templates part of a release, are templates _for_ the specific release. And as @rmohr pointed out above, they are not portable between releases. Templates - manifests - need to be considered to be an artifact of the release itself. This is important to understand.

Nevertheless even this per-release templates should have a significant value, as - as noted above - they are parametric.

**Comment by rmohr** [comment]
Resolved by #1116. First release with a published template is this: https://github.com/kubevirt/kubevirt/releases/tag/v0.7.0-alpha.5
